filechooser-portal/Main.vala

```diff
@@ -343,10 +339,58 @@ public class Files.FileChooserPortal : Object {
         yield;
 
         dialogs.remove (parent_window);
+        if (dialog.register_id != 0) {
+            connection.unregister_object (dialog.register_id);
+        }
+
+        dialog.dispose ();
         response = _response;
         results = _results;
     }
 
+    private Gtk.ResponseType check_overwrite (GLib.File? file) {
+        if (file == null) {
+            return Gtk.ResponseType.CANCEL;
+        }
+
+        if (file.query_exists ()) {
+            var display_name = file.get_basename ();
+            var primary = _("Replace XXXX “%s”?").printf (display_name);
+            unowned var secondary = _("Replacing this file will overwrite its current contents");
+            if (file.query_file_type (FileQueryInfoFlags.NOFOLLOW_SYMLINKS) == FileType.SYMBOLIC_LINK) {
+                try {
+                    var info = file.query_info (FileAttribute.STANDARD_SYMLINK_TARGET, FileQueryInfoFlags.NONE);
+                    if (info != null) {
+                        primary = _("Replace “%s”?".printf (info.get_symlink_target ()));
+                    } else {
+                        primary = _("Replace the target of “%s”?".printf (display_name));
+                    }
+
+                    secondary = _("Replacing the target file for this link will overwrite its current contents.");
+                } catch (Error e) {
+                    warning ("Could not get info for %s", file.get_uri ());
+                }
+            }
+
+            var replace_dialog = new Granite.MessageDialog.with_image_from_icon_name (
+                primary, secondary, "dialog-warning", Gtk.ButtonsType.CANCEL
+            );
+
+            var replace_button = replace_dialog.add_button ("Replace", Gtk.ResponseType.YES);
+            replace_button.get_style_context ().add_class (Gtk.STYLE_CLASS_DESTRUCTIVE_ACTION);
+
+            var replace_response = replace_dialog.run ();
```

Comment:
`Gtk.Dialog.run()` was also removed in Gtk4, to block until the response, you need an `new MainLoop(null, true)` and call `MainLoop.quit()` in the `Gtk.Dialog.response()` signal.

Reply:
Rather than messing with MainLoop I have made the overwrite check async with a modal dialog.  Seems to work - let me know if there are flaws.